DM-52719: Use UUIDv7 for dataset IDs

dhirving opened the pull request:
Generate v7 UUIDs for dataset IDs instead of v4 UUIDs.  v4 UUIDs create problems for Postgres insert performance, because they  scatter index updates randomly around the disk.  v7 UUIDs have a timestamp in the most significant bits, so datasets generated at the same time will be close together in the index.  Initial testing shows up to 100x improvement in insert speed with v7 UUIDs.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Use UUIDv7 for dataset IDs

Generate v7 UUIDs for dataset IDs instead of v4 UUIDs.  v4 UUIDs create problems for Postgres insert performance, because they  scatter index updates randomly around the disk.  v7 UUIDs have a timestamp in the most significant bits, so datasets generated at the same time will be close together in the index.  Initial testing shows up to 100x improvement in insert speed with v7 UUIDs.

TallJimbo commented:
Looks great, works great.

Note that we actually generate most of our UUIDs in practice in `pipe_base`, since that happens before we ever make a `DatasetRef` from them.  I've got some commits on pipe_base for utilizing this code there, and if you don't mind I might just make a pipe_base branch for this ticket.  It's not just a simple swap since I actually just merged some code that makes  UUID4 assumptions to speed up some lookups, but I've already got a fix.

In light of that, should we move this to a more public module in daf_butler?

dhirving commented:
> Looks great, works great.
> 
> Note that we actually generate most of our UUIDs in practice in `pipe_base`, since that happens before we ever make a `DatasetRef` from them. I've got some commits on pipe_base for utilizing this code there, and if you don't mind I might just make a pipe_base branch for this ticket. It's not just a simple swap since I actually just merged some code that makes UUID4 assumptions to speed up some lookups, but I've already got a fix.
> 
> In light of that, should we move this to a more public module in daf_butler?

Sure, I'll move it to the `_rubin` "internal public" package.

timj commented:
Are we intending to prefer the python 3.14 implementation if available? I assume the UUIDs in pipe_base are the quanta UUIDs and not the DatasetRef ones.

dhirving commented:
> Are we intending to prefer the python 3.14 implementation if available? I assume the UUIDs in pipe_base are the quanta UUIDs and not the DatasetRef ones.

I think we should just have one implementation -- why introduce variable behavior when we don't have to.  (The standard library is currently using a more complex variant that provides monotonicity in the generated IDs that we don't need.)   When we no longer have to support 3.13 we can get rid of this code and switch to the standard library version.

dhirving committed:
Export generate_uuidv7 from _rubin

Mark generate_uuidv7 as "internal-public" by exporting it from the _rubin package.

TallJimbo commented:
`pipe_base` is generating both dataset and quantum UUIDs in the QG builder; its internal data structures don't use `DatasetRef` until well after it needs the IDs.
